databricks-sdk-java/src/main/java/com/databricks/sdk/core/ApiClient.java

```diff
@@ -29,55 +30,119 @@
+    private RetryStrategyPicker retryStrategyPicker;
+    private boolean isDebugHeaders;
+
+    public Builder withDatabricksConfig(DatabricksConfig config) {
+      this.authenticateFunc = v -> config.authenticate();
+      this.getHostFunc = v -> config.getHost();
+      this.getAuthTypeFunc = v -> config.getAuthType();
+      this.httpClient = config.getHttpClient();
+      this.debugTruncateBytes = config.getDebugTruncateBytes();
+      this.accountId = config.getAccountId();
+      this.retryStrategyPicker = new RequestBasedRetryStrategyPicker(config.getHost());
+      this.isDebugHeaders = config.isDebugHeaders();
+
+      return this;
+    }
+
+    public Builder withTimer(Timer timer) {
+      this.timer = timer;
+      return this;
+    }
+
+    public Builder withAuthenticateFunc(Function<Void, Map<String, String>> authenticateFunc) {
+      this.authenticateFunc = authenticateFunc;
+      return this;
+    }
+
+    public Builder withGetHostFunc(Function<Void, String> getHostFunc) {
+      this.getHostFunc = getHostFunc;
+      return this;
+    }
+
+    public Builder withGetAuthTypeFunc(Function<Void, String> getAuthTypeFunc) {
+      this.getAuthTypeFunc = getAuthTypeFunc;
+      return this;
+    }
+
+    public Builder withHttpClient(HttpClient httpClient) {
+      this.httpClient = httpClient;
+      return this;
+    }
+
+    public Builder withRetryStrategyPicker(RetryStrategyPicker retryStrategyPicker) {
+      this.retryStrategyPicker = retryStrategyPicker;
+      return this;
+    }
+
+    public ApiClient build() {
+      return new ApiClient(this);
+    }
+  }
+
   private static final Logger LOG = LoggerFactory.getLogger(ApiClient.class);
 
   private final int maxAttempts;
 
   private final ObjectMapper mapper;
 
-  private final DatabricksConfig config;
-
   private final Random random;
 
   private final HttpClient httpClient;
   private final BodyLogger bodyLogger;
   private final RetryStrategyPicker retryStrategyPicker;
   private final Timer timer;
+  private final Function<Void, Map<String, String>> authenticateFunc;
+  private final Function<Void, String> getHostFunc;
+  private final Function<Void, String> getAuthTypeFunc;
+  private final String accountId;
+  private final boolean isDebugHeaders;
   private static final String RETRY_AFTER_HEADER = "retry-after";
 
   public ApiClient() {
     this(ConfigLoader.getDefault());
   }
 
   public String configuredAccountID() {
-    return config.getAccountId();
+    return accountId;
   }
 
   public ApiClient(DatabricksConfig config) {
     this(config, new SystemTimer());
   }
 
   public ApiClient(DatabricksConfig config, Timer timer) {
-    this.config = config;
-    config.resolve();
+    this(new Builder().withDatabricksConfig(config.resolve()).withTimer(timer));
+  }
 
-    Integer rateLimit = config.getRateLimit();
-    if (rateLimit == null) {
-      rateLimit = 15;
-    }
+  private ApiClient(Builder builder) {
+    this.timer = builder.timer;
```

Comment:
Suggestion: let's use sane defaults when those values are not provided in the builder. Ideally `ApiClient.Builder().build()` would make a more or less working ApiClient. Here, we can use `SystemTimer()` if not specified.